### PR TITLE
Game logic

### PR DESCRIPTION
**Pull request by eypsrcnuygr**
* This is the branch for the game logic.
* Whether you can play with someone or with CPU.
* We put the winning conditions.

### CODE REVIEW

**Comment by eypsrcnuygr** [comment]
* Separate the logic from the interface.
* More descriptive namings.

**Comment by sumancrest0001** [comment]
**Status: Changes required :gear: 
Hello TEAM,** 
You have done a good job and thank you for adding the CPU option also in the project. Your project is working correctly. But we have to follow the best practices while writing code. For example, you are using classes to only initialize symbols and board. We have to follow OOP and taking advantage of it.

[Required] Third step of the getting started section of the readme file throws an error when I follow the instruction.

[Optional] When we use CPU as second player, sometimes we get incorrect input messages multiple times as you are using the **random method** to select a cell. For the CPU player, we can use the minimax algorithm, watch [this](https://www.youtube.com/watch?v=trKjYdBASyQ&vl=en) video if you want to learn more.

**Comment by defoebrand** [comment]
Best practices regarding what should be in a class seems to vary between TSEs, as the first TSE to review our project uses a class that only exists to initialize variables.
Readme instructions are not part of this step in the milestone, please reserve comments on readme files for that step.
Minmax looks like an interesting option and we will explore this before finalizing our project in the last milestone

File: lib/game.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+class Game
+  attr_accessor :choices, :board
+  def initialize
```
Comment (Dandush03) [inline lib/game.rb:3]: ### Use `attr_accessor` ONLY when needed.

File: lib/player_o.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class PlayerO
+  attr_accessor :symbol
```
Comment (Dandush03) [inline lib/player_o.rb:2]: ### Use `attr_accessor` ONLY when needed.

File: lib/player_x.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class PlayerX
+  attr_accessor :symbol
```
Comment (Dandush03) [inline lib/player_x.rb:2]: ### Use `attr_accessor` ONLY when needed.

File: lib/winning_condition.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+  def inside(board)
+    board[1] == board[4] && board[1] == board[7] && board[1] != '_' ||
+      board[3] == board[4] && board[3] == board[5] && board[3] != '_'
+  end
+
+  def diagonal(board)
+    board[2] == board[4] && board[2] == board[6] && board[2] != '_' ||
+      board[0] == board[4] && board[0] == board[8] && board[0] != '_'
+  end
```
Comment (Dandush03) [inline lib/winning_condition.rb:24]: ### Avoid using repeated code, try using a loop for validation.

File: bin/main.rb [hunk 5]
```diff
@@ -1,60 +1,122 @@
...
+require_relative '../lib/game'
+require_relative '../lib/winning_condition'
+require_relative '../lib/player_x'
+require_relative '../lib/player_o'
 
-        puts 'Incorrect input'
-        player1
-      end
+def player1(choices, board)
+  puts 'Player X - choose square 1-9'
```
[the remark this answers is not in the document]
Comment (eypsrcnuygr) [reply bin/main.rb:9]: Done

File: bin/main.rb [hunk 6]
```diff
@@ -1,60 +1,122 @@
...
+    puts 'Incorrect input'
+    player1(choices, board)
   end
+end
 
-  def player2
-    puts 'Player O - choose square 1-9'
-    o_input = gets.chomp
-    if o_input.match?(/\b[1-9]\b/)
+def player2(choices, board, opponent)
```
[the remark this answers is not in the document]
Comment (eypsrcnuygr) [reply bin/main.rb:25]: Done

File: bin/main.rb [hunk 7]
```diff
@@ -1,60 +1,122 @@
...
+    p 'Goodbye!'
+    exit
+  end
+end
 
-        puts 'Incorrect input'
-        player2
-      end
+def choose_opponent(opponent)
+  if opponent.nil?
```
[the remark this answers is not in the document]
Comment (eypsrcnuygr) [reply bin/main.rb]: Done

File: lib/game.rb (hunk 1 above)
Comment (eypsrcnuygr) [reply lib/game.rb:3]: Done

File: lib/player_o.rb (hunk 2 above)
Comment (eypsrcnuygr) [reply lib/player_o.rb]: Done

**Comment by uryelah** [comment]
Hi **Sercan Uygur & Defoe Defoe**, I'm Sarah your reviewer for this PR 👩‍🚀 

Good job with the changes, the difference is visible in your code.

Now, while you still have some logic in your bin folder that could easily be refactored into the Game class you seem to have grasped the main OOP principles so I'll leave further refactoring as an [OPTIONAL] exercise. Think about which methods really need to be in the main file, try to make it so you only call methods and print string in the main file, the leaner the better as all needed features to play the game should ideally be contained inside the game classes.

All is fine and you are good to go! 🚀 

-------

## **Your project is approved! 🎊 🥅 💛**

**You may close this Pull Request and merge the changes**


---------

    
    
**Keep being awesome devs and good luck with your next project!**

---------

Cheers and Happy coding!👏👏👏

Code review by **Sarah Uryelah Silva Chamorro**

(Feel free to contact me on slack by the handle of  **sarah** if you have any additional question)

**Pull request merged**